### ISSUE TITLE
Enhancement #1: Ensure FlowUnit Derivatives Match Exact Mathematical Derivatives (Close to PyTorch's Implementation)

### ISSUE THREAD

**Issue by Shyanil** [Shyanil/HyperFlow#2]
#### **Description**  
The current implementation of `FlowUnit` derivatives follows **real mathematical differentiation exactly**, meaning it correctly applies the derivative rules. However, in some cases, mathematical derivatives **ignore negligible values**, while PyTorch’s autograd system uses a **high-level computational model** that may handle these cases differently.  

This enhancement aims to:
- **Keep the derivatives strictly aligned with real mathematical differentiation.**
- **Ensure the computed gradients remain close to PyTorch's autograd system**, while recognizing that PyTorch may use approximations or different numerical techniques.
- **Improve the precision and consistency** of backpropagation calculations.

3. **Expected Improvement**  
   - Gradients should be **mathematically exact**, avoiding minor computational discrepancies.
   - The values should be **closer to PyTorch’s autograd results**, without deviating from true differentiation rules.

#### **Next Steps**  
- **Refine the backpropagation logic** to ensure precise gradient computation.
- **Compare results with PyTorch autograd** for verification while keeping mathematical correctness.
- **Test multiple edge cases** to confirm accuracy.

This enhancement will improve the reliability of gradient computations while maintaining compatibility with mathematical principles and PyTorch’s behavior.

---

```python
# Test the FlowUnit class functions
def test_flow_unit_functions():
    # Create a FlowUnit instance
    data = np.array([1.0, 2.0, 3.0])
    flow_unit = FlowUnit(data)

    # Test sigmoid
    sigmoid_out = flow_unit.sigmoid()
    print("Sigmoid Output:", sigmoid_out.data)
    sigmoid_out.backpropagate()
    print("Sigmoid Gradient:", flow_unit.grad)

    # Test tanh
    tanh_out = flow_unit.tanh()
    print("Tanh Output:", tanh_out.data)
    tanh_out.backpropagate()
    print("Tanh Gradient:", flow_unit.grad)

    # Test relu
    relu_out = flow_unit.relu()
    print("ReLU Output:", relu_out.data)
    relu_out.backpropagate()
    print("ReLU Gradient:", flow_unit.grad)

    # Test leaky_relu
    leaky_relu_out = flow_unit.leaky_relu(alpha=0.01)
    print("Leaky ReLU Output:", leaky_relu_out.data)
    leaky_relu_out.backpropagate()
    print("Leaky ReLU Gradient:", flow_unit.grad)

    # Test softmax
    softmax_out = flow_unit.softmax()
    print("Softmax Output:", softmax_out.data)
    softmax_out.backpropagate()
    print("Softmax Gradient:", flow_unit.grad)

# Test the LossFunctions class functions
def test_loss_functions():
    # Create FlowUnit instances for logits and target
    logits = FlowUnit(np.array([2.0, 1.0, 0.1]))
    target = [1, 0, 0]  # One-hot encoded target

    # Test categorical_cross_entropy
    cce_loss = LossFunctions.categorical_cross_entropy(logits, target)
    print("Categorical Cross-Entropy Loss:", cce_loss.data)
    cce_loss.backpropagate()
    print("Categorical Cross-Entropy Gradient:", logits.grad)

    # Create FlowUnit instances for inputs and target
    inputs = FlowUnit(np.array([[1.0, 2.0], [3.0, 4.0]]))
    target = FlowUnit(np.array([1, 0]))
    parameters = (FlowUnit(np.array([0.5, -0.5])), FlowUnit(0.1))

    # Test binary_cross_entropy_loss
    bce_loss = LossFunctions.binary_cross_entropy_loss(inputs, target, parameters)
    print("Binary Cross-Entropy Loss:", bce_loss.data)
    bce_loss.backpropagate()
    print("Binary Cross-Entropy Gradient:", inputs.grad)
    
    # Test mse_loss
    mse_loss = LossFunctions.mse_loss(inputs, target, parameters)
    print("Mean Squared Error Loss:", mse_loss.data)
    mse_loss.backpropagate()
    print("Mean Squared Error Gradient:", inputs.grad)

# Run the tests
if __name__ == "__main__":
    test_flow_unit_functions()
    test_loss_functions()
```

#### **Sample Output:**
```
Sigmoid Output: [0.73105858 0.88079708 0.95257413]
Sigmoid Gradient: [0.19661193 0.10499359 0.04517666]
Tanh Output: [0.76159416 0.96402758 0.99505475]
Tanh Gradient: [0.61658627 0.17564441 0.0550427 ]
ReLU Output: [1. 2. 3.]
ReLU Gradient: [1.61658627 1.17564441 1.0550427 ]
Leaky ReLU Output: [1. 2. 3.]
Leaky ReLU Gradient: [2.61658627 2.17564441 2.0550427 ]
Softmax Output: [0.09003057317038046, 0.24472847105479764, 0.6652409557748218]
Softmax Gradient: [1.3877787807814457e-17, 2.7755575615628914e-17, 8.326672684688674e-17]
Categorical Cross-Entropy Loss: 0.4170300011033529
Categorical Cross-Entropy Gradient: 0.0
Binary Cross-Entropy Loss: [0.71561525]
Binary Cross-Entropy Gradient: 0.0
Mean Squared Error Loss: [0.655]
Mean Squared Error Gradient: 0.0
